### DISCUSSION TITLE
Custom nodejs-buildpacks failed to build the image

### DISCUSSION THREAD

**Discussion by Sunny-001** [paketo-buildpacks/feedback#51] [General]
I changed the nodejs download address to my private server and executed "./scripts/package.sh --version 0.5" in the build folder to generate the following file

![image](https://user-images.githubusercontent.com/46052043/201676215-934fdf80-87ac-47dc-ba28-34140e66da12.png)

then used the following command to build the image,but the process had an error

pack build registry.cn-hangzhou.aliyuncs.com/kpack/ruoyi-ui:2.0.0 --builder registry.cn-hangzhou.aliyuncs.com/kpack/paketo-buildpacks-builder:base -b build/buildpack.tgz --env "BP_NODE_RUN_SCRIPTS=build:prod"   --env "NODE_ENV=development"   --pull-policy if-not-present --path /data/build-package-sample/buildpack-test/ruoyi-ui
===> DETECTING
**[detector] ERROR: No buildpack groups passed detection.
[detector] ERROR: Please check that you are running against the correct path.
[detector] ERROR: failed to detect: no buildpacks participating
ERROR: failed to build: executing lifecycle. This may be the result of using an untrusted builder: failed with status code: 20**

**Comment by sophiewigmore** [comment]
Hi @Sunny-001! I think detection is failing because you're using a builder, which has pre-defined buildpack groups, but then you're also defining a buildpack to use. By passing in a single buildpack, you're overriding those groups and ONLY providing a single buildpack to the build.

If you want to pass in a custom buildpack, you need to pass in all of the buildpacks in the group.

**Reply by sophiewigmore** [reply]
For example (if your app uses Node Engine,  NPM Install, and Node Start)
```
pack build registry.cn-hangzhou.aliyuncs.com/kpack/ruoyi-ui:2.0.0 --builder registry.cn-hangzhou.aliyuncs.com/kpack/paketo-buildpacks-builder:base -b build/buildpack.tgz -b paketo-buildpacks/npm-install -b paketo-buildpacks/node-start
```

**Comment by sophiewigmore** [comment]
Separately, you might consider using a [dependency mapping](https://paketo.io/docs/howto/configuration/#dependency-mappings) type [service binding](https://paketo.io/docs/howto/configuration/#bindings) instead to use your own dependency. If you use a dependency mapping for your dependency, then you don't even need to pass in your own buildpack, you can just do something like this, where the binding is a dependency mapping (from the service binding docs):

```
pack build --volume <absolute-path-to-binding>:/platform/bindings/<binding-name> <image-name>
```

**Reply by bitgully** [reply]
Hi @sophiewigmore and @dmikusa!
Do you have a solution for using dependency mappings that do not reference the local file system but a private registry that requires authentication (e.g. HTTP Basic)? The download (via a proxy) would have to pull the artefact from https://private-registry.foo.bar but include an Authorization header in doing so.

**Reply by dmikusa** [reply]
Dependency mappings don't have to point to the local file system they can point anywhere, so it can reference a URL. You can put a user/password into the URL and I think it might use that as HTTP basic credentials. I haven't tried it, but if it doesn't work it should, so you could open an issue against https://github.com/paketo-buildpacks/libpak/issues. I can't speak for buildpacks based on packit, maybe @sophiewigmore knows.

**Reply by bitgully** [reply]
Thanks @dmikusa. Putting username/password in the URL works. But those credentials get printed to stdout in plain text at the moment. I have opened a PR to solve this issue: https://github.com/paketo-buildpacks/libpak/pull/312

**Reply by dmikusa** [reply]
Awesome! Thanks

**Comment by Sunny-001** [comment]
I want to download Node Engine from my private server. What should I do

**Reply by sophiewigmore** [reply]
You can either do what you are doing with your custom buildpack or follow the dependency mapping comment I posted above.

> If a dependency URI is inaccessible from the build environment, a [binding](https://paketo.io/docs/howto/configuration/#bindings) can be used to map a new URI to a given dependency. This allows organizations to upload a copies of vetted dependencies to an accessible location and provide developers and CI/CD pipelines with configuration pointing the buildpack at the accessible dependencies.

**Reply by Sunny-001** [reply]
The information you provided is very useful, thank you very much!

**Comment by Sunny-001** [comment]
I found some useful information [here](https://paketo.io/docs/howto/configuration/#dependency-mappings), maybe I see what you mean, I will try it tomorrow

**Comment by dmikusa** [comment]
In regards to dependency mapping, I typically find those work best for local development. It can be tricky to get them into a build system/CI/kpack. 

They can also be a little tricky to get formatted correctly, which is why I created the `binding-tool` cli, or `bt`. [This SO post](https://stackoverflow.com/a/74406677/1585136) has some information the tool and how it can make generating the files easier. It works very well with any Paketo buildpack, just point the tool at a buildpack, it'll download the deps and generate the files.

**Reply by Sunny-001** [reply]
Thank you very much,I will try tomorrow

**Reply by Sunny-001** [reply]
@dmikusa First of all, thank you very much for your help, I learned a lot of useful information from your reply and applied it to the build process, and it was very useful. But I ran into a problem when building the vue project. The build command is as follows:

`pack build ruoyi-ui:2.0.0 \
  --path ./RuoYi-Vue/ \
  --builder paketobuildpacks/builder:base \
  --env "BP_NODE_RUN_SCRIPTS=build:prod" \
  --env "NODE_ENV=development" \
  --env "BP_NODE_PROJECT_PATH=ruoyi-ui" \
  --env SERVICE_BINDING_ROOT=/bindings \
  --volume /root/npm-bindings/:/bindings/npmrc \
  --pull-policy if-not-present`

I created the corresponding file according to the instructions, set the environment variable SERVICE_BINDING_ROOT during construction, and mounted the directory /root/npm-bindings/, the directory structure and file contents of /root/npm-bindings/ are as follows：

![image](https://user-images.githubusercontent.com/46052043/201863275-11407c6b-4dd2-496b-b368-977ea887bb77.png)

this configuration does not appear to be in effect, and the following error occurs at build time
![image](https://user-images.githubusercontent.com/46052043/201863634-eb4ac2d2-b6fa-4a98-9f3b-9323d9ca62f6.png)

If I place the.npmrc file directly in the vue project directory, you can clearly see the build speed
![image](https://user-images.githubusercontent.com/46052043/201864147-6da4e5ee-f7ef-47e3-830c-479ce5b1397e.png)
So, is there a problem with the configuration in the build command or the configuration file

**Reply by dmikusa** [reply]
Sorry, that's getting out of my area of expertise, @paketo-buildpacks/nodejs-maintainers - any ideas on the above?

**Comment by Sunny-001** [comment]
@sophiewigmore Can you help me with this question? 
![image](https://user-images.githubusercontent.com/46052043/201956573-246b71c2-4d58-454d-912c-f56384cb7dcc.png)

**Reply by ryanmoran** [reply]
Do you see a line that says `Loading service binding of type 'npmrc'` in your build log output?

**Reply by Sunny-001** [reply]
the command and build log are as follows, `Loading service binding of type 'npmrc'` did not appear.

-bash-4.2# pack build ruoyi-ui:2.0.0 \
>   --path ./RuoYi-Vue/ \
>   --builder paketobuildpacks/builder:base \
>   --env "BP_NODE_RUN_SCRIPTS=build:prod" \
>   --env "NODE_ENV=development" \
>   --env "BP_NODE_PROJECT_PATH=ruoyi-ui" \
>   --env SERVICE_BINDING_ROOT=/bindings \
>   --volume /root/npm-bindings/:/bindings/npmrc \
>   --pull-policy if-not-present
===> DETECTING
6 of 10 buildpacks participating
paketo-buildpacks/ca-certificates 3.0.1
paketo-buildpacks/node-engine     0.11.2
paketo-buildpacks/npm-install     0.6.2
paketo-buildpacks/node-module-bom 0.2.0
paketo-buildpacks/node-run-script 0.3.0
paketo-buildpacks/npm-start       0.6.1
===> ANALYZING
Restoring metadata for "paketo-buildpacks/ca-certificates:helper" from app image
Restoring metadata for "paketo-buildpacks/node-engine:node" from app image
Restoring metadata for "paketo-buildpacks/npm-install:modules" from app image
Restoring metadata for "paketo-buildpacks/npm-install:npm-cache" from app image
Restoring metadata for "paketo-buildpacks/node-module-bom:cyclonedx-node-module" from cache
===> RESTORING
Restoring data for "paketo-buildpacks/node-engine:node" from cache
Restoring data for "paketo-buildpacks/npm-install:modules" from cache
Restoring data for "paketo-buildpacks/npm-install:npm-cache" from cache
Restoring data for "paketo-buildpacks/node-module-bom:cyclonedx-node-module" from cache
===> BUILDING

Paketo CA Certificates Buildpack 3.0.1
  https://github.com/paketo-buildpacks/ca-certificates
  Launch Helper: Reusing cached layer
Paketo Node Engine Buildpack 0.11.2
  Resolving Node Engine version
    Candidate version sources (in priority order):
      package.json -> "14.18.1"
      <unknown>    -> ""

    Selected Node Engine version (using package.json): 14.18.1

  Reusing cached layer /layers/paketo-buildpacks_node-engine/node

Paketo NPM Install Buildpack 0.6.2
  Resolving installation process
    Process inputs:
      node_modules      -> "Not found"
      npm-cache         -> "Not found"
      package-lock.json -> "Not found"

    Selected NPM build process: 'npm install'

  Executing build process
    Running 'npm install --unsafe-perm --cache /layers/paketo-buildpacks_npm-install/npm-cache'
npm install failed: exit status 1
    npm ERR! code ECONNREFUSED
    npm ERR! errno ECONNREFUSED
    npm ERR! FetchError: request to https://registry.npmjs.org/@vue%2fcli-plugin-babel failed, reason: connect ECONNREFUSED 104.16.22.35:443
    npm ERR!     at ClientRequest.<anonymous> (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/node-fetch-npm/src/index.js:68:14)
    npm ERR!     at ClientRequest.emit (events.js:400:28)
    npm ERR!     at TLSSocket.socketErrorListener (_http_client.js:475:9)
    npm ERR!     at TLSSocket.emit (events.js:400:28)
    npm ERR!     at emitErrorNT (internal/streams/destroy.js:106:8)
    npm ERR!     at emitErrorCloseNT (internal/streams/destroy.js:74:3)
    npm ERR!     at processTicksAndRejections (internal/process/task_queues.js:82:21)
    npm ERR! From previous event:
    npm ERR!     at cachingFetch (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/make-fetch-happen/index.js:126:47)
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/npm-registry-fetch/index.js:88:50
    npm ERR!     at processImmediate (internal/timers.js:464:21)
    npm ERR! From previous event:
    npm ERR!     at regFetch (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/npm-registry-fetch/index.js:88:37)
    npm ERR!     at fetchPackument (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/lib/fetchers/registry/packument.js:42:10)
    npm ERR!     at packument (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/lib/fetchers/registry/packument.js:20:10)
    npm ERR!     at getManifest (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/lib/fetchers/registry/manifest.js:22:10)
    npm ERR!     at manifest (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/lib/fetchers/registry/manifest.js:13:10)
    npm ERR!     at Object.manifest (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/lib/fetchers/registry/index.js:17:12)
    npm ERR!     at Object.Fetcher#manifest [as manifest] (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/genfun/lib/genfun.js:15:38)
    npm ERR!     at manifest (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/lib/fetch.js:23:18)
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/manifest.js:24:12
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/promise-inflight/inflight.js:29:24
    npm ERR! From previous event:
    npm ERR!     at _inflight (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/promise-inflight/inflight.js:28:25)
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/promise-inflight/inflight.js:22:14
    npm ERR!     at processImmediate (internal/timers.js:464:21)
    npm ERR! From previous event:
    npm ERR!     at inflight (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/promise-inflight/inflight.js:14:40)
    npm ERR!     at Object.manifest (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/manifest.js:22:10)
    npm ERR!     at fetchPackageMetadata (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/lib/fetch-package-metadata.js:62:10)
    npm ERR!     at callFunc (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/call-limit/call-limit.js:25:12)
    npm ERR!     at limited (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/call-limit/call-limit.js:13:7)
    npm ERR!     at addDependency (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/lib/install/deps.js:599:5)
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/lib/install/deps.js:539:5
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/async-map.js:52:35
    npm ERR!     at Array.forEach (<anonymous>)
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/async-map.js:52:11
    npm ERR!     at Array.forEach (<anonymous>)
    npm ERR!     at asyncMap (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/async-map.js:51:8)
    npm ERR!     at exports.loadDevDeps (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/lib/install/deps.js:533:3)
    npm ERR!     at Array.<anonymous> (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/bind-actor.js:15:8)
    npm ERR!     at LOOP (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/chain.js:15:14)
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/chain.js:18:7
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/lib/install/and-finish-tracker.js:8:8
    npm ERR!     at zalgoSafe (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/dezalgo/dezalgo.js:20:10)
    npm ERR!     at LOOP (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/chain.js:7:26)
    npm ERR!  FetchError: request to https://registry.npmjs.org/@vue%2fcli-plugin-babel failed, reason: connect ECONNREFUSED 104.16.22.35:443
    npm ERR!     at ClientRequest.<anonymous> (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/node-fetch-npm/src/index.js:68:14)
    npm ERR!     at ClientRequest.emit (events.js:400:28)
    npm ERR!     at TLSSocket.socketErrorListener (_http_client.js:475:9)
    npm ERR!     at TLSSocket.emit (events.js:400:28)
    npm ERR!     at emitErrorNT (internal/streams/destroy.js:106:8)
    npm ERR!     at emitErrorCloseNT (internal/streams/destroy.js:74:3)
    npm ERR!     at processTicksAndRejections (internal/process/task_queues.js:82:21)
    npm ERR! From previous event:
    npm ERR!     at cachingFetch (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/make-fetch-happen/index.js:126:47)
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/npm-registry-fetch/index.js:88:50
    npm ERR!     at processImmediate (internal/timers.js:464:21)
    npm ERR! From previous event:
    npm ERR!     at regFetch (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/npm-registry-fetch/index.js:88:37)
    npm ERR!     at fetchPackument (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/lib/fetchers/registry/packument.js:42:10)
    npm ERR!     at packument (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/lib/fetchers/registry/packument.js:20:10)
    npm ERR!     at getManifest (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/lib/fetchers/registry/manifest.js:22:10)
    npm ERR!     at manifest (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/lib/fetchers/registry/manifest.js:13:10)
    npm ERR!     at Object.manifest (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/lib/fetchers/registry/index.js:17:12)
    npm ERR!     at Object.Fetcher#manifest [as manifest] (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/genfun/lib/genfun.js:15:38)
    npm ERR!     at manifest (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/lib/fetch.js:23:18)
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/manifest.js:24:12
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/promise-inflight/inflight.js:29:24
    npm ERR! From previous event:
    npm ERR!     at _inflight (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/promise-inflight/inflight.js:28:25)
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/promise-inflight/inflight.js:22:14
    npm ERR!     at processImmediate (internal/timers.js:464:21)
    npm ERR! From previous event:
    npm ERR!     at inflight (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/promise-inflight/inflight.js:14:40)
    npm ERR!     at Object.manifest (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/pacote/manifest.js:22:10)
    npm ERR!     at fetchPackageMetadata (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/lib/fetch-package-metadata.js:62:10)
    npm ERR!     at callFunc (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/call-limit/call-limit.js:25:12)
    npm ERR!     at limited (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/call-limit/call-limit.js:13:7)
    npm ERR!     at addDependency (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/lib/install/deps.js:599:5)
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/lib/install/deps.js:539:5
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/async-map.js:52:35
    npm ERR!     at Array.forEach (<anonymous>)
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/async-map.js:52:11
    npm ERR!     at Array.forEach (<anonymous>)
    npm ERR!     at asyncMap (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/async-map.js:51:8)
    npm ERR!     at exports.loadDevDeps (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/lib/install/deps.js:533:3)
    npm ERR!     at Array.<anonymous> (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/bind-actor.js:15:8)
    npm ERR!     at LOOP (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/chain.js:15:14)
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/chain.js:18:7
    npm ERR!     at /layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/lib/install/and-finish-tracker.js:8:8
    npm ERR!     at zalgoSafe (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/dezalgo/dezalgo.js:20:10)
    npm ERR!     at LOOP (/layers/paketo-buildpacks_node-engine/node/lib/node_modules/npm/node_modules/slide/lib/chain.js:7:26) {
    npm ERR!   type: 'system',
    npm ERR!   errno: 'ECONNREFUSED',
    npm ERR!   code: 'ECONNREFUSED',
    npm ERR!   parent: 'ruoyi-ui'
    npm ERR! }
    npm ERR!
    npm ERR! If you are behind a proxy, please make sure that the
    npm ERR! 'proxy' config is set properly.  See: 'npm help config'

    npm ERR! A complete log of this run can be found in:
    npm ERR!     /layers/paketo-buildpacks_npm-install/npm-cache/_logs/2022-11-16T00_53_13_139Z-debug.log

ERROR: failed to build: exit status 1
ERROR: failed to build: executing lifecycle: failed with status code: 51

**Reply by fg-j** [reply]
It looks like you're using a very old version of the buildpacks! For instance, the latest version of the npm-install buildpack is [0.10.8](https://github.com/paketo-buildpacks/npm-install/releases/tag/v0.10.8) and you're using 0.6.2. Your trouble with bindings may be caused by [a bug related to bindings detection](https://github.com/paketo-buildpacks/packit/issues/275) that we patched in subsequent versions. Try upgrading to the latest versions of the constituent buildpacks and let us know if your problem persists!

**Reply by Sunny-001** [reply]
How can I specify the version?

**Reply by Sunny-001** [reply]
![D803ED2F-57CD-4628-843B-382585BD227A](https://user-images.githubusercontent.com/46052043/202234653-b2d97ce4-f453-48fa-a34d-f32ee239e7ac.png)

Just like this?

**Comment by robdimsdale** [comment]
Yes, you could use `-b` multiple times to manually add each of the buildpacks. If you do this, you will need to specific each and every buildpack - it creates a new build order which much have a complete set of buildpacks in it.

However, the `builder` you are using contains the buildpacks within it. So if you update your builder to the latest version, you will get the latest buildpacks. That's simpler and quicker than using `-b` for each buildpack.

**Reply by Sunny-001** [reply]
ok, I will try tomorrow

**Reply by Sunny-001** [reply]
![image](https://user-images.githubusercontent.com/46052043/202335341-803f6acc-3ea7-4a96-a4e8-0aa9ac4d53e8.png)

What is the tag of the latest paketobuildpacks/builder image? I tried latest, but npm-install buildpack is not the latest version

**Reply by Sunny-001** [reply]
I tried both the `0.3.103-base` and `latest` versions and got an error:
`ERROR: failed to build: validating stack mixins: index.docker.io/paketobuildpacks/run:base-cnb missing required mixin(s): libexpat1`

**Reply by fg-j** [reply]
Your issue is that you are setting `--pull-policy=if-not-present`. That means `pack` won't pull a newer version of the builder if you've already got one on your local docker daemon. Try removing that flag. You'll also probably want to use the full builder for your app -- at least at first. You can specify with `--builder paketobuildpacks/builder:full`. With those two suggestions together, you should end up getting the latest buildpacks and the latest full stack.

**Reply by robdimsdale** [reply]
Oh I totally missed the `--pull-policy` flag @fg-j ! Thanks for catching it. @Sunny-001 let us know if this resolves your issue.

**Comment by robdimsdale** [comment]
Yeah that's odd. Unfortunately, I'm not familiar enough with the concept of `mixins` to know why this is failing with that error. @ryanmoran @paketo-buildpacks/stacks-maintainers have you seen this sort of issue before?

Some additional context:

The current npm-install buildpack is [v0.10.8](https://github.com/paketo-buildpacks/npm-install/releases/tag/v0.10.8). And the `0.3.103-base` contains that version of npm-install (see e.g. https://github.com/paketo-buildpacks/base-builder/releases/tag/v0.3.103). So, you probably want to stick with this builder once we resolve the mixin issu.

I also notice that your output contains: `restoring ... from cache`. That might be irrelevant, but you can use `--clear-cache` to ensure that any stale, cached, build data is ignored.

Do you possibly have some sort of other config - e.g a [project.toml file](https://buildpacks.io/docs/app-developer-guide/using-project-descriptor/)? That might be irrelevant too, I just want to get a clear picture of what your app looks like.

**Reply by Sunny-001** [reply]
yes, i see the version of npm-install buildpack in `paketobuildpacks/builder:0.3.103-base` is `0.10.8`, but the process shown the npm-install buildpack version is 0.4.0.

**Reply by robdimsdale** [reply]
I agree. That's very confusing. That's why I was wondering if you have some sort of additional config that's overriding the `--builder` flag.

A couple of other things to check:

* what version of `pack` are you running
* is there any possibility that your connection to dockerhub is being cached (e.g. by a corporate proxy) and returning a stale version of `paketobuildpacks/builder:base`?
* What happens if you try `--builder gcr.io/paketo-buildpacks/builder:0.3.103-base` instead? It's the same image as what is hosted on dockerhub.

**Reply by robdimsdale** [reply]
Also, is it possible for you to share your app somewhere so we can try it out?

**Reply by Sunny-001** [reply]
the pack version as follows:
![image](https://user-images.githubusercontent.com/46052043/202350830-a911cefd-6889-42a2-a822-7f29afb40849.png)

I am in China and the image`gcr.io/ paceto -buildpacks/builder:0.3.103-base` I cannot access.
I'm using an open source project, and here's the project address: [https://github.com/Sunny-001/RuoYi-Vue.git](url)

**Reply by fg-j** [reply]
See my [comment](https://github.com/orgs/paketo-buildpacks/discussions/51#discussioncomment-4168483) about resolving this issue!

**Reply by Sunny-001** [reply]
@robdimsdale @fg-j thank you very much! I have solved the problem, my docker is configured with `"registry-mirrors":["xxxxxx"]`, i has deleted this configuration , reload docker daemon and restart docker, now, the .npmrc configuration file can be load. but I don't know why the configuration `"registry-mirrors":["xxxxxx"]`  caused this problem

**Reply by Sunny-001** [reply]
The environment variable `BP_NODE_PROJECT_PATH` does not work in `0.3.103-base` and `full` versions of builder `paketo-buildpacks/builder`

**Reply by fg-j** [reply]
@Sunny-001 Are you describing a new problem you're encountering re: `BP_NODE_PROJECT_PATH`?

**Reply by Sunny-001** [reply]
yes

**Reply by fg-j** [reply]
To keep things organized, can you please [file an issue](https://github.com/paketo-buildpacks/nodejs/issues/new/choose) on the Nodejs buildpack repository? Adding some reproduction steps and full log output on your issue will help the Nodejs maintainers resolve your issue.

**Reply by Sunny-001** [reply]
18/5000 
I have created a new [issue](https://github.com/paketo-buildpacks/nodejs/issues/683) for this problem
